### PR TITLE
handle missing instance_type in reserved pricing

### PR DESCRIPTION
**Pull request by stanmluke**
handle missing instance_type in reserved pricing, when it does not exist in non-reserved; also bump amazon-pricing version gem.

the fetch_ec2_instance_pricing_ri_v2() assumes that all instance_types are populated by the non-reserved call fetch_ec2_instance_pricing(), which is _not_ true for x1.32xlarge.  (I believe the on-demand case is handled by "reserved hosts".)

Assume bumping gem version is appropriate here, so we can capture this new functionality, probably should've bumped in the previous submit.

Unsure about Gemfile.lock changes, so calling that out here.

These changes get me past a sandbox collection test.

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -1,20 +1,17 @@
...
-    amazon-pricing (0.1.76)
+    amazon-pricing (0.1.77)
 
 GEM
   remote: https://rubygems.org/
   specs:
-    coderay (1.1.0)
+    coderay (1.1.1)
     diff-lcs (1.1.3)
+    ffi (1.9.10-java)
```
Comment (lardcanoe) [inline Gemfile.lock]: java stuff? hmmmm... not sure i am good with that being added even though harmless.

**Comment by jeffzhou791** [comment]
I don't quite get the "(I believe the on-demand case is handled by "reserved hosts)" part. Did you check if your change impact other EC2 prices?

**Comment by stanmluke** [comment]
@jeffzhou791  I meant "dedicated hosts", and not "reserved hosts";  sorry.

> Did you check if your change impact other EC2 prices?

yes.  The only instance_type which is added (as a result of this change) is "x1.32xlarge".  The bug here was insidious, since it only occurred in this special case;  i.e. that AWS would have a 1:1 correspondence of instance-types in ondemand and reserved pricing.
So the reserved-instance expects all instance_types to be populated by the ondemand results (Ec2DiPriceList::get_ec2_di_od_pricing);  and this was not the case for reserved-dedicated (Ec2DiPriceList::get_ec2_reserved_di_pricing).

Note that there is no issue with Ec2PriceList::get_ec2_on_demand_instance_pricing and Ec2PriceList::get_ec2_reserved_instance_pricing.  Because all reserved instance_type appear in both;  rather than continue to make this assumption, I allow an instance_type to occur only in reserved.

We can see this issue on this AWS page: [dedicated-instances](https://aws.amazon.com/ec2/purchasing-options/dedicated-instances/).  Notice there is no "Dedicated On-Demand" pricing for x1.32xlarge, and I verified by directly downloading the pricing page.  

There is a related link on the above AWS page [compare Dedicated Instances and Dedicated Hosts](https://aws.amazon.com/ec2/dedicated-hosts/), which further describes dedicated hosts.

We don't seem to incorporate dedicated-hosts into our amazon-pricing model, which is a hole in our product?

File: lib/amazon-pricing/common/ec2_common.rb [hunk 2]
```diff
@@ -41,14 +41,38 @@ def fetch_ec2_instance_pricing_ri_v2(url, operating_system)
           next
         end
         reg['instanceTypes'].each do |type|
-          api_name = type["type"]
-          instance_type = region.get_instance_type(api_name)
+          # it's possible we're missing the instance_type (i.e. it wasn't in the base, cf: ec2-di-price-list),
```
Comment (jeffzhou791) [inline lib/amazon-pricing/common/ec2_common.rb]: can I suggest a more readable organization of code:

reg['instanceTypes'].each do |type|
 ensure_existence_of_instance_type type
  ...
end

def ensure_existence_of_instance_type type
  # Create instance type
  instance_type = find_or_create_instance_type type
  # Fill in od price if missing
  fill_od_price_if_missing instance_type
end

def find_or_create_instance_type type
...
end

def fill_od_price_if_missing instance_type
...
end

File: lib/amazon-pricing/definitions/ec2-instance-type.rb [hunk 3]
```diff
@@ -98,6 +98,24 @@ def operating_systems
...
+      if not throughput
+        logger.warn "Unknown network throughput for instance type #{api_name}"
+      end
+      throughput
+    end
+    def self.get_network_mbps(throughput)
+      if throughput
+        network_mbps = @Network_Throughput_Bytes_Per_Hour[throughput]
+      else
+        logger.warn "Unknown network throughput #{throughput}"
```
Comment (jeffzhou791) [inline lib/amazon-pricing/definitions/ec2-instance-type.rb]: nit: if we were following the same pattern, this logger.warn will be on '!network_mbps', not on 'throughput'?

File: lib/amazon-pricing/definitions/ec2-instance-type.rb [hunk 4]
```diff
@@ -114,5 +132,130 @@ def self.get_name(instance_type, api_name, is_reserved = false)
...
+        :very_low => 50,    # FIXME check this value
+        :low => 100,
+        :low_to_moderate => 250,  # FIXME check this value
+        :moderate => 500,
+        :high => 1000,
+        :ten_gigabit => 10000
+    }
+
+    @Network_Performance = {
+      'c1.medium' => :moderate,
```
Comment (jeffzhou791) [inline lib/amazon-pricing/definitions/ec2-instance-type.rb]: thanks for getting this list.

File: lib/amazon-pricing/definitions/ec2-instance-type.rb [hunk 5]
```diff
@@ -114,5 +132,130 @@ def self.get_name(instance_type, api_name, is_reserved = false)
...
     end
 
+    # throughput values for performance ratings
+    # Amazon informally tells customers that they translate as follows:
+    #   Low = 100 Mbps
+    #   Medium = 500 Mbps
+    #   High = 1 Gbps
+    #   10Gig = 10 Gbps (obviously)
+    @Network_Throughput_Bytes_Per_Hour = {
+        :very_low => 50,    # FIXME check this value
```
Comment (jeffzhou791) [inline lib/amazon-pricing/definitions/ec2-instance-type.rb]: what needs to be done here? how do we get throughput number for :very_low today?

**Comment by jeffzhou791** [comment]
👍

**Pull request merged**